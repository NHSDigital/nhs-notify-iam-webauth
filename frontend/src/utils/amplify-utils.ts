/* eslint-disable import/no-unresolved,@typescript-eslint/no-require-imports */
/* eslint-disable unicorn/prefer-module */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import { cookies } from 'next/headers';
import { createServerRunner } from '@aws-amplify/adapter-nextjs';
import { fetchAuthSession } from 'aws-amplify/auth/server';

const config = require('@/amplify_outputs.json');

export const { runWithAmplifyServerContext } = createServerRunner({
  config,
});

export async function getAccessTokenServer(): Promise<string | undefined> {
  try {
    const { tokens } = await runWithAmplifyServerContext({
      nextServerContext: { cookies },
      operation: fetchAuthSession,
    });

    return tokens?.accessToken?.toString();
  } catch {
    // no-op
  }
}
