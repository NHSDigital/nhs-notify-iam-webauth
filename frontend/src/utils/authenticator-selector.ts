import { useAuthenticator } from '@aws-amplify/ui-react';

type UseAuthenticatorSelector = Exclude<
  Parameters<typeof useAuthenticator>[0],
  undefined
>;

export const authenticatorSelector: UseAuthenticatorSelector = (ctx) => [
  ctx.authStatus,
];
