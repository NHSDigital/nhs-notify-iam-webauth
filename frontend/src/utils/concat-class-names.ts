type ClassName = string | undefined | false;

export default function concatClassNames(...classNames: ClassName[]): string {
  return classNames
    .filter(Boolean)
    .join(' ')
    .replaceAll(/\s{2,}/g, ' ')
    .trim();
}
