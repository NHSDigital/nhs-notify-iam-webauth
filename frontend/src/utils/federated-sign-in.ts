import { signInWithRedirect } from '@aws-amplify/auth';
// eslint-disable-next-line import-x/no-unresolved
import amplifyOutputs from '@/amplify_outputs.json';

export interface State {
  redirectPath: string;
}

export function federatedSignIn(redirectPath: string) {
  const providers = amplifyOutputs.auth?.oauth?.identity_providers || [];
  if (providers.length !== 1) {
    throw new Error('Invalid OAUTH provider configuration');
  }
  const provider = providers[0];
  if (!provider) {
    throw new Error('Missing OAUTH custom provider configuration');
  }
  return signInWithRedirect({
    provider: {
      custom: provider,
    },
    customState: JSON.stringify({ redirectPath }),
  });
}
