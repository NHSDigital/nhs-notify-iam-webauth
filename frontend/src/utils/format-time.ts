export const formatTime = (remainingSeconds: number) => {
  const remainingMinutes = Math.round(remainingSeconds / 60);

  if (remainingMinutes <= 0) {
    return `${remainingSeconds} seconds`;
  }

  if (remainingMinutes <= 1) {
    return '1 minute';
  }

  return `${remainingMinutes} minutes`;
};
