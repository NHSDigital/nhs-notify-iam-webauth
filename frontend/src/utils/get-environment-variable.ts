export const getEnvironmentVariable = (name: string) => {
  const secret = process.env[name];

  if (!secret) {
    throw new Error('Missing environment variable');
  }

  return secret;
};
