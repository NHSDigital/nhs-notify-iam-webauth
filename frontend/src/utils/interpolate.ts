/**
 * Replaces variables within a template string using a variables object, and supports pluralisation.
 *
 * Supports:
 * - simple variable interpolation: `{{name}}`
 * - pluralisation: `{{variable|singular|plural}}`
 *
 * Example:
 * interpolate('Hello {{name}}, you have {{count}} {{count|message|messages}}.', {
 *   name: 'John',
 *   count: 2,
 * }); // => 'Hello John, you have 2 messages.'
 */

// eslint-disable-next-line security/detect-unsafe-regex, sonarjs/slow-regex
const interpolationPattern = /{{([^|}]+)(?:\|([^|}]+)\|([^}]+))?}}/g;

export function interpolate(
  template: string,
  variables: Record<string, string | number> = {}
): string {
  // eslint-disable-next-line unicorn/prefer-string-replace-all
  return template.replace(interpolationPattern, (_, key, singular, plural) => {
    const value = variables[key];

    if (singular !== undefined && plural !== undefined) {
      const numeric = Number(value);

      if (Number.isNaN(numeric)) return plural;
      return numeric === 1 ? singular : plural;
    }

    return value == null ? '' : String(value);
  });
}
