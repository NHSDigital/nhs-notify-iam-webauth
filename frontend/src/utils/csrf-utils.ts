'use server';

import { createHmac, randomBytes } from 'node:crypto';
import { cookies } from 'next/headers';
import { logger } from '@nhs-notify-iam-webauth/utils-logger';
import { getEnvironmentVariable } from './get-environment-variable';
import { getSessionId } from './amplify-utils';

export const generateSessionCsrfToken = async (sessionId: string) => {
  const secret = getEnvironmentVariable('CSRF_SECRET');

  const salt = randomBytes(8).toString('hex');

  const hash = createHmac('sha256', secret)
    .update(sessionId)
    .update(salt)
    .digest('hex');

  const csrfToken = `${hash}.${salt}`;

  return csrfToken;
};

export const verifyCsrfToken = async (
  csrfToken: string,
  secret: string,
  sessionId: string
) => {
  const [hmac, salt] = csrfToken.split('.');

  const expectedHmac = createHmac('sha256', secret)
    .update(sessionId)
    .update(salt)
    .digest('hex');

  return expectedHmac === hmac;
};

export const verifyFormCsrfToken = async (formData: FormData) => {
  const secret = getEnvironmentVariable('CSRF_SECRET');
  const sessionId = await getSessionId();

  if (!sessionId) {
    logger.error('Unauthenticated');
    return false;
  }

  const cookieStore = await cookies();
  const csrfTokenCookie = cookieStore.get('csrf_token');

  if (!csrfTokenCookie) {
    logger.error('missing CSRF cookie');
    return false;
  }

  const csrfTokenCookieValue = csrfTokenCookie.value;

  const csrfTokenFormField = formData.get('csrf_token');

  if (!csrfTokenFormField) {
    logger.error('missing CSRF form field');
    return false;
  }

  if (csrfTokenFormField !== csrfTokenCookieValue) {
    logger.error('CSRF mismatch');
    return false;
  }

  const formVerification = await verifyCsrfToken(
    csrfTokenFormField,
    secret,
    sessionId
  );

  if (!formVerification) {
    logger.error('CSRF error');
    return false;
  }

  return true;
};
