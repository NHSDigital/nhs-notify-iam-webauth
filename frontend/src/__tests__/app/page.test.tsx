import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UseAuthenticator, useAuthenticator } from '@aws-amplify/ui-react';
import { redirect, RedirectType } from 'next/navigation';
import { mockDeep } from 'jest-mock-extended';
import { federatedSignIn } from '@/src/utils/federated-sign-in';
import SignInPage from '@/src/app/page';

const mockGetSearchParams = jest.fn();

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  useSearchParams: () => ({
    get: mockGetSearchParams,
  }),
  redirect: jest.fn(),
}));

jest.mock('@aws-amplify/ui-react', () => ({
  Authenticator: () => <p>Placeholder Sign-in form</p>,
  useAuthenticator: jest.fn(() =>
    mockDeep<UseAuthenticator>({ authStatus: 'unauthenticated' })
  ),
}));

jest.mock('@/src/components/CIS2SignInButton/CIS2SignInButton', () => ({
  CIS2SignInButton: ({ onClick }: { onClick: () => void }) => (
    <button type='button' data-testid='mock-cis2-button' onClick={onClick}>
      Mock CIS2 Sign In Button
    </button>
  ),
}));

jest.mock('@/src/utils/federated-sign-in');

const mockFederatedSignIn = jest.mocked(federatedSignIn);
const mockUseAuthenticator = jest.mocked(useAuthenticator);
const mockRedirect = jest.mocked(redirect);

describe('SignInPage', () => {
  const originalCognitoIdpSetting = process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP;

  it('renders', async () => {
    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = 'false';

    const container = render(<SignInPage />);

    expect(container.asFragment()).toMatchSnapshot();

    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = originalCognitoIdpSetting;
  });

  it('renders with cognito login form if env var switch is enabled', () => {
    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = 'true';

    const container = render(<SignInPage />);

    expect(container.asFragment()).toMatchSnapshot();

    process.env.NEXT_PUBLIC_ENABLE_COGNITO_IDP = originalCognitoIdpSetting;
  });

  it('redirects using search parameter if already signed in', () => {
    mockUseAuthenticator.mockReturnValueOnce(
      mockDeep<UseAuthenticator>({ authStatus: 'authenticated' })
    );

    mockGetSearchParams.mockReturnValueOnce('/example-redirect');

    render(<SignInPage />);

    expect(mockRedirect).toHaveBeenCalledWith(
      '/signin?redirect=%2Fexample-redirect',
      RedirectType.push
    );
  });

  it('redirects to /templates/manage-templates if already signed in and there is no redirect search parameter', () => {
    mockUseAuthenticator.mockReturnValueOnce(
      mockDeep<UseAuthenticator>({ authStatus: 'authenticated' })
    );

    render(<SignInPage />);

    expect(mockRedirect).toHaveBeenCalledWith(
      '/signin?redirect=%2Ftemplates%2Fmanage-templates',
      RedirectType.push
    );
  });

  describe('CIS2 login', () => {
    it('does federated sign in when clicking cis2 button', async () => {
      const user = userEvent.setup();

      const container = render(<SignInPage />);

      const button = container.getByTestId('mock-cis2-button');

      await user.click(button);

      expect(mockFederatedSignIn).toHaveBeenCalledWith(
        '/templates/manage-templates'
      );
    });

    it('sets redirect based on search params when clicking cis2 button', async () => {
      const user = userEvent.setup();

      mockGetSearchParams.mockReturnValueOnce('/example-redirect');

      const container = render(<SignInPage />);

      const button = container.getByTestId('mock-cis2-button');

      await user.click(button);

      expect(mockFederatedSignIn).toHaveBeenCalledWith('/example-redirect');
    });
  });
});
