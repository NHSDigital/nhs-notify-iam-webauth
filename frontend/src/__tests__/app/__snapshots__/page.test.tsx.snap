// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInPage renders 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        class="nhsuk-heading-xl"
      >
        Sign in
      </h1>
      <div
        class="nhsuk-u-padding-6 notify-content"
      >
        <h2
          class="nhsuk-heading-m"
        >
          Sign in using an NHS account
        </h2>
        <button
          data-testid="mock-cis2-button"
          type="button"
        >
          Mock CIS2 Sign In Button
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SignInPage renders with cognito login form if env var switch is enabled 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1
        class="nhsuk-heading-xl"
      >
        Sign in
      </h1>
      <div
        class="nhsuk-u-padding-6 notify-content"
      >
        <h2
          class="nhsuk-heading-m"
        >
          Sign in using an NHS account
        </h2>
        <button
          data-testid="mock-cis2-button"
          type="button"
        >
          Mock CIS2 Sign In Button
        </button>
      </div>
      <hr
        class="nhsuk-section-break nhsuk-section-break--m"
      />
      <div
        class="nhsuk-u-padding-6 notify-content notify-cognito-sign-in-form"
      >
        <h2
          class="nhsuk-heading-m"
        >
          Dev Only - Sign in with Email / Password
        </h2>
        <p>
          Placeholder Sign-in form
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
